RDF::IndexingService indexes objects & properties

awead opened the pull request:
Solrizing AF::Base objects and properties share common behaviors. This refactors both methods for indexing objects and properties into a common method: `solrize_rdf_assertions` which creates other classes that can be overridden if users wish to alter the ways in which their solr fields are created.

This also addresses #809 where objects can be indexed using their uris as the value.

awead commented:
@terrellt yo :dog2:, I hear you like indexing.

terrellt commented:
@awead Did you put indexing in my indexing?

terrellt commented on lib/active_fedora/rdf/indexing_service.rb:
```diff
@@ -64,35 +64,107 @@ def resource
-        config = properties[name.to_s]
-        config && config[:class_name] && config[:class_name] < ActiveFedora::Base
+      # Override this method to use your own FieldMap class for custom indexing of objects and properties
+      def field_map_class
```
I'm welcome to be overridden, but this is a textbook factory. `field_map_factory`?

jcoyne replied:
Doesn't this just give you the problem of "Which factory class should I instantiate?" `FactoryFactory`?

terrellt replied:
Not sure I understand

jcoyne replied:
Neither do I. Can you give an example of what you want it changed to?

terrellt commented on lib/active_fedora/rdf/indexing_service.rb:
```diff
@@ -64,35 +64,107 @@ def resource
+
+        def insert(name, index_field_config, object)
+          self[index_field_config.key.to_s] ||= FieldMapEntry.new
+          builder = kind_of_af_base?(name, object) ? resource_builder_class : property_builder_class
```
Can you extract this logic to a polymorphic builder factory?

terrellt replied:
To clarify, I'm imagining something like this...
```ruby
class PolymorphicBuilder
  def build(entry, index_field_config, object, name)
    delegate_class(name, object).build(entry, index_field_config, object, name)
  end

  private

  def delegate_class(name, object)
    if kind_of_af_base?(name, object)
      ResourceBuilder
    else
      PropertyBuilder
    end
  end

  def kind_of_af_base(name, object)
    config = properties(object)[name.to_s]
    config && config[:class_name] && config[:class_name] < ActiveFedora::Base
  end

  def properties(object)
    object.class.properties
  end
end
```

When you write it out this way you start to see some patterns emerging too - object/name are passed around a lot, it seems like those may be something you'd want in a constructor.

terrellt commented:
There's a lot of nested value objects in the rdf indexing service which are potentially useful in other contexts. Can they be extracted?

awead commented:
@terrellt I'm not sure what you mean about the factory stuff, but I took your PolymorphicBuilder and refactored with that and split out the classes to their own files.

terrellt commented:
@awead It was just a method name comment, but we can leave it as is for consistency with the rest of the code base. I had one last note on the polymorphic builder, but after that discussion and potential changes I'm :+1:

terrellt commented:
Oh, I lied. Documentation comments incoming.

terrellt commented on lib/active_fedora/rdf/field_map.rb:
```diff
@@ -0,0 +1,80 @@
+      yield self if block_given?
+    end
+
+    def insert(name, index_field_config, object)
```
Documentation please.

terrellt commented on lib/active_fedora/rdf/field_map.rb:
```diff
@@ -0,0 +1,80 @@
+
+    # Abstract class for building FieldMap entries
+    class Builder
+      def self.build(entry, index_field_config, object, name)
```
Documentation please.

terrellt commented on lib/active_fedora/rdf/field_map_entry.rb:
```diff
@@ -0,0 +1,24 @@
+      @values = []
+    end
+
+    def merge!(type, behaviors, values)
```
Documentation please.

awead commented:
@terrellt add the documentation. I also smoothed out the PolymorphicBuild/Builder interface so that they now are the same. You can also chain the methods now, like: `PolymorphicBuilder.new(a,b,c,d).build`
